enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,249 @@
+    portTarget:
+      fixedIPs:
+      - subnet:
+          id: d8350f6a-b6d8-48d8-b0e2-7a5ff8da6353
+      - subnet:
+          id: 29ea3065-c47c-475e-8f63-19548a30a2ac
+      network:
+        id: c6ccc703-819b-4c27-9cf0-657ab6d0f2c7
+```
+
+4. The customer will generate the manifests and configured MachineConfigs
+for the master and workers depending on the IPv6 addresses mode chosen for the Subnets
+in step 1. If SLAAC or Stateless was used the MachineConfigs has to contain the followning:
+
+```yaml
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: master
+  name: 05-master-eui64
+spec:
+  config:
+    ignition:
+      version: 3.2.0
+    storage:
+      files:
+      - contents:
+          # [connection]
+          # type=ethernet
+          # [ipv6]
+          # addr-gen-mode=eui64
+          # method=auto
+          source: data:text/plain;charset=utf-8;base64,W2Nvbm5lY3Rpb25dCnR5cGU9ZXRoZXJuZXQKW2lwdjZdCmFkZHItZ2VuLW1vZGU9ZXVpNjQKbWV0aG9kPWF1dG8=
+        filesystem: root
+        mode: 0600
+        overwrite: true
+        path: /etc/NetworkManager/system-connections/nmconnection.template
+```
+
+While when using stateful it has to contain the following in the kernel arguments:
+
+```yaml
+kind: MachineConfig
+metadata:
+  labels:
+    machineconfiguration.openshift.io/role: master
+  name: 05-master-kernelarg-dhcp
+spec:
+  config:
+    ignition:
+      version: 3.2.0
+  kernelArguments:
+    - ip=dhcp,dhcp6
+```
+6. The CIDRs of the subnets have to added to the `cloud-provider-config.yaml`
+in the `address-sort-order` field to make sure IPv4 address will have higher priority in the
+Node addresses.
+7. The installer will validate the subnets and VIPs, add the security group to the api and ingress
+ports and link the floating IPs to those Ports. Also, create the additional IPv6 security group rules
+and configured the servers with dual-stack addresses, add the dual-stack allowed address pairs and
+include in the bootstrap node the following Network Manager configuration:
+
+```ini
+[connection]
+type=ethernet
+[ipv6]
+method=auto
+```
+8. In machine-config-operator, kubelet will be update to not be configured with node-ips.
+9. cloud-provider-openStack will filter the node addresses that are meant to be specified in the Status of the Node
+by looking for a tag that was applied in the main Ports of the Node.
+
+### API Extensions
+
+The `machinesSubnet` field, which is available in the `install-config.yaml` under
+`platform` and `openstack` will be deprecated in favor of a new field named `portTarget`
+to persist both IPv4 and IPv6 provided Subnets.
+
+For installation on OpenStack with dual-stack kubelet will no longer rely on the node-ip to get the node
+addresses, given that when using external cloud provider, the node-ip currently has no support for dual-stack, this [enhancement](https://github.com/kubernetes/enhancements/pull/3706) describes it. Instead,
```

Comment:
Given that the upstream enhancement has been approved and implemented and will be available in OpenShift by the time this one is implemented, does that change your approach? The one-off thing we did for vsphere was intended to be a temporary solution until proper dual stack cloud provider support was available. Since it now is, you don't necessarily need to do the intermediate step (unless there are other reasons for this not listed here).

Reply:
I think there's an implicit assumption here that --node-ip is a good way forward. My own view is that this should be knowable in the cluster explicitly, without relying on the runtime output of a heuristic script executed on the node itself.

My current preference for how to achieve this is to explicitly indicate interfaces which should be used for control plane traffic (i.e. exposed in Node.Addresses). This way the cloud provider can set them explicitly without requiring kubelet to tell it which interfaces to set. This makes more sense to me on multiple levels:

Firstly, kubelet telling cloud-provider which node addresses to tell kubelet it should use was only ever a workaround.

Secondly, this is fundamentally a decision made during provisioning. The simplest and most robust solution is to store this decision and use it directly.

Reply:
>Firstly, kubelet telling cloud-provider which node addresses to tell kubelet it should use was only ever a workaround.

I agree, if we draw a graph of the flow it's indeed ugly, but...

>without relying on the runtime output of a heuristic script executed on the node itself

... even today with using `--node-ip` no one is forced to use heuristics and you have ability to say "I want X and Y to be Node IPs". If you say this, no heuristic will be ran and you will end up with what you want - a deterministic way of defining what lands in Node.Status.Addresses

Reply:
In OpenShift, `--node-ip` is set by the output of a script[^1]. This is the heuristics I'm referring to. Sure you can set `--node-ip`, but something needs to know what to set it to. In my view the ultimate source of truth is whatever provisioned the machine in the first place. Something somewhere added this network interface with the intention that it would be used for handling internal control plane traffic. I would like that information to flow robustly and unambiguously from the provisioner to where it's used.

It could even be used to set `--node-ip`, tbh. For e.g. in OpenStack we could add some server metadata identifying the control plane interface, and then we could replace the linked script with something that knows precisely what was intended without guesswork. I think it's probably easier for the CCM to do it, though.

[^1]: I think it's this one, but I haven't checked properly: https://github.com/openshift/baremetal-runtimecfg/blob/master/scripts/nodeip-finder

Reply:
https://github.com/openshift/baremetal-runtimecfg/blob/master/cmd/runtimecfg/node-ip.go is the correct one, not the Python one, but that's a detail.

The main point is, I agree that for OpenStack (and other platforms with CCM) it's a smart move to let whoever created the network interface also pass the information that this is the interface that should be used internally for the control traffic. I am not saying that if we have this information, we should ignore it.

But I am saying, that the opposite flow is not scaling across all the platforms we support in OpenShift. I do not want to simply drop baremetal-runtimecfg for an arbitrary platform X, because platform X has CCM that is the ultimate oracle. Because next door you have platform Y which has CCM without this knowledge, and next next door you have platform Z which doesn't have CCM at all.

The ideal flow for me is that baremetal-runtimecfg listens to whoever comes with authoritative info. So I would like your CCM to simply say "please use THIS interface because I say so" and baremetal-runtimecfg simply respect it (no heuristics). This way we keep design of all the on-prem platforms the same -- otherwise we are quickly running into scenario when every platform uses different set of components "just because".

And we already have a proven track of this kind of a wishlist inside baremetal-runtimecfg, so I can simply say that you are not the first one with such a request and we already have a partial implementation for it (partial, because the previous person asking didn't have CCM in his platform).

Reply:
What environment-independent mechanism that baremetal-runtimecfg could consume unambiguously directly would you recommend? I can't think of anything myself, and I suspect we'd end up doing environment-specific on the node in order to create something that baremetal-runtimecfg could then consume, e.g. retrieving information from a local metadata service. At that point it feels simpler to just make the intended change directly.

Reply:
we have decided to rely on `CloudDualStackNodeIPs` feature gate instead. We revisit the way on how the node addresses are handled on a separate enhancement.